src/tools/cgexec.c

```diff
@@ -171,12 +188,128 @@ int main(int argc, char *argv[])
 		}
 	}
 
-	/* Now exec the new process */
-	ret = execvp(argv[optind], &argv[optind]);
-	if (ret == -1) {
-		err("%s", strerror(errno));
+	if (!replace_idle) {
+		/* Now exec the new process */
+		ret = execvp(argv[optind], &argv[optind]);
+		err("exec failed:%s", strerror(errno));
+		return ret;
+	}
+
+	/*
+	 * By now the cgexec is migrated to the right cgroup(s),
+	 * 1. fork() and in the child path perform the exec to program
+	 *    passed in the command line.  This means, we have cgexec
+	 *    and program executing in the desired cgroup(s).
+	 * 2. In the parent, parse /proc/<pid>/cgroup of the current
+	 *    process and if the cgroup name ends with .scope and has
+	 *    a task with /proc/<pid>/cmdline as
+	 *    "libcgroup_systemd_idle_thread", kill it.  This gives the
+	 *    illusion of replacement.
+	 */
+	exec_task_pid = fork();
+	if (exec_task_pid == -1) {
+		err("Fork failed:%s\n", strerror(errno));
+		return  -1;
+	}
+
+	/* child process exec's into task program */
+	if (exec_task_pid == 0) {
+		execvp(argv[optind], &argv[optind]);
+		err("exec failed:%s", strerror(errno));
+		return  -1;
+	}
+
+	snprintf(pid_proc_cgrp_path, FILENAME_MAX, "/proc/%u/cgroup", pid);
+	pid_proc_cgrp_fp = fopen(pid_proc_cgrp_path, "re");
+	if (!pid_proc_cgrp_fp) {
+		err("Failed to open: %s\n", pid_proc_cgrp_path);
 		return -1;
 	}
 
+	while (fgets(buffer, FILENAME_MAX, pid_proc_cgrp_fp)) {
+		char ctrl_name[CONTROL_NAMELEN_MAX];
+		char *ctrl;
+		int len;
+
+		/* read according to the cgroup mode */
+		if (strstr(buffer, "::"))
+			ret = sscanf(buffer, "%d::%4096s\n", &id, cgroup_name);
+		else
+			ret = sscanf(buffer, "%d:%[^:]:%4096s\n", &id, ctrl_name, cgroup_name);
+
+		/* skip named controllers */
+		if (ret == 3 && strncmp(ctrl_name, "name", 4) == 0)
+			continue;
+
+		/* skip if the cgroup doesn't ends with .scope */
+		base = basename(cgroup_name);
+		len = strlen(base) - 6;
+		if (strcmp(base + len, ".scope") != 0)
+			continue;
+
+		/* cgroup v2 (unified) mode */
+		if (ret == 2) {
+			scope_pid = find_scope_idle_task_pid(cgroup_name, NULL);
+		} else {
+			/* cgroup v1 (legacy/hybrid) mode */
+			/* check if the mount point is shared between controllers */
+			ctrl = strchr(ctrl_name, ',');
+			if (ctrl) {
+				/* truncate other controllers, but first */
+				len = strlen(ctrl_name) - strlen(ctrl);
```

Comment:
This looks like it works, but wouldn't it be simpler to just do:
`ctrl[0] = '\0';`

Reply:
It's trying to extract the first controller. for example, `cpu,cpuacct`, truncate `,cpuacct`